lens_shading_analyse.c

```diff
@@ -316,27 +316,48 @@ int main(int argc, char *argv[])
 		uint16_t *channel = out_buf[channel_ordering[bayer_order][i]];
 		int mid_value_avg = 0;
 		int count = 0;
-		//Average out the centre 64 pixels.
-		for (x=(single_channel_width>>1)-4; x<=(single_channel_width>>1)+4; x++)
-		{
-			for (y=(single_channel_height>>1)-4; y<=(single_channel_height>>1)+4; y++)
-			{
-				mid_value_avg += channel[x + y*single_channel_width];
-				count++;
-			}
-		}
-		uint16_t middle_val = (mid_value_avg / count) << 5;
-
+		uint16_t *line;
 		const char *channel_comments[4] = {
 			"R",
 			"Gr",
 			"Gb",
 			"B"
-		};
-		printf("Middle_val is %d\n", middle_val);
+		};	
+		
+		// Search highest value
+		float max_val = 0;
+		for (y=0; y<grid_height; y++)
+		{
+			int y_start = y*32;
+			int y_stop  = y_start + 32 <= single_channel_height ? y_start + 32 : single_channel_height;
+			for (x=0; x<grid_width; x++)
+			{
+				int x_start = x*32;
+				int x_stop  = x_start + 32 <= single_channel_width ? x_start + 32 : single_channel_width;
 
+				float    block_val = 0;
+				uint16_t block_px = 0;
+
+				for(int y_px = y_start; y_px < y_stop; y_px++){
+					line = &channel[y_px*(single_channel_width)];
+					for(int x_px = x_start; x_px < x_stop; x_px++){
+						block_val += line[x_px];
+						block_px++;
+					}
+				}
+
+				block_val /= block_px;
+				if(block_val > max_val){
+					max_val = block_val;
+				}
+			}
+		}
+
+		max_val *= 32;
+		printf("Max_val is %.0f\n", max_val);
```

Comment:
Without a better description I can't say for certain what this is meant to be doing, but it's very ugly to run the same loop twice, once to find this max_val, and then again to compute the cell gains.

Allocate arrays for block_val and block_px (if it's really necessary) and compute once. Calculating the gains then becomes an iteration over those arrays instead of the whole image.

(Sorry, embedded programming always makes you think of more efficient code).

Reply:
Running the loop twice is of course not a high-performance solution. But it runs very fast on my Raspberry Pi 3B ;-) 
Quick and dirty - I will fix it.

Reply:
Double calculation loop removed